### PR TITLE
DRILL-3171: handle concurrent Zookeeper node creation gracefully

### PR DESCRIPTION
**Pull request by hnfgns**
Make ZK node creation/updates logic optimistic.
Refactor ZkAbstractStore to eliminate redundant code around node creation logic.
Enhanced documentation.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java [hunk 1]
```diff
@@ -23,30 +23,35 @@
...
 import java.util.Map.Entry;
+import java.util.concurrent.TimeUnit;
 
 import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.api.ACLBackgroundPathAndBytesable;
 import org.apache.curator.framework.recipes.cache.ChildData;
 import org.apache.curator.framework.recipes.cache.PathChildrenCache;
 import org.apache.curator.framework.recipes.cache.PathChildrenCache.StartMode;
-import org.apache.drill.exec.rpc.data.DataTunnel;
+import org.apache.curator.framework.recipes.locks.InterProcessLock;
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java]: unused imports.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java [hunk 2]
```diff
@@ -23,30 +23,35 @@
...
  * This is the abstract class that is shared by ZkPStore (Persistent store) and ZkEStore (Ephemeral Store)
  * @param <V>
  */
 public abstract class ZkAbstractStore<V> implements AutoCloseable {
-  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ZkAbstractStore.class);
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ZkAbstractStore.class);
 
-  protected CuratorFramework framework;
-  protected PStoreConfig<V> config;
+  protected static final long LOCK_TIMEOUT_MILLIS = 3000;
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java]: unused.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java [hunk 3]
```diff
@@ -54,31 +59,26 @@ public ZkAbstractStore(CuratorFramework framework, PStoreConfig<V> config)
     this.prefix = parent + "/";
     this.framework = framework;
     this.config = config;
+    this.childrenCache = new PathChildrenCache(framework, parent, true);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java]: unnecessary "this"

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java [hunk 4]
```diff
@@ -136,7 +138,81 @@ public boolean putIfAbsent(String key, V value) {
...
+        if (isUpdate) {
+          creator.forPath(path, valueInBytes);
+        } else {
+          creator.forPath(path);
+        }
+      } else if (isUpdate) {
+        framework.setData().forPath(path, valueInBytes);
+      }
+    } catch (KeeperException.NodeExistsException ex) {
+      logger.warn("Node already exists in Zookeeper. Skipping... -- [path: {}, mode: {}]", path, mode);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java]: If isUpdate == true, should we try to update?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkPStore.java [hunk 5]
```diff
@@ -18,8 +18,11 @@
 package org.apache.drill.exec.store.sys.zk;
 
 import java.io.IOException;
+import java.util.concurrent.TimeUnit;
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkPStore.java]: unused imports.

**Comment by sudheeshkatkam** [comment]
How does this address the problem mentioned in the JIRA, which seems to come from **StoragePluginMap**?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java (hunk 4 above)
Comment (hnfgns) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java]: We already do handle two cases? i) nodeExists & update ii) !nodeExists & update.

**Comment by hnfgns** [comment]
I will look into StoragePluginMap as well.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java (hunk 3 above)
Comment (hnfgns) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkAbstractStore.java]: I consistently use `this` (i) in c'tor assignments or (ii) if there is any shadowing variable.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkEStore.java [hunk 6]
```diff
@@ -17,40 +17,25 @@
...
-  public ZkEStore(CuratorFramework framework, PStoreConfig<V> config) throws IOException{
+  public ZkEStore(CuratorFramework framework, PStoreConfig<V> config) throws IOException {
     super(framework,config);
   }
 
   @Override
-  public void delete(String key) {
-    try {
-      if (framework.checkExists().forPath(p(key)) != null) {
-        framework.delete().forPath(p(key));
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkEStore.java]: Previously the node was not rebuilt for ZkEStore (see `childrenCache.rebuildNode(withPrefix(key))` in ZkAbstractStore). Is this change necessary?

Comment (hnfgns) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/zk/ZkEStore.java]: This ensures that the cache is updated after deletion takes place.

**Comment by sudheeshkatkam** [comment]
LGTM.

**Comment by sudheeshkatkam** [comment]
+1 ;)
